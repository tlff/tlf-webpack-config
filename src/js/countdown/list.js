import baseClass from "../lib/event";
import utils from "../utils";
import "icheck";
import "icheck/skins/flat/red.css";
import $ from 'jquery';
import datatable from "datatables";
$.fn.DataTable =datatable(window,$);
export default class countList extends baseClass {
    constructor(id) {
        super();
        if (typeof id == "object") {
            this.container = id;
        } else {
            this.container = $("#" + id);
        }
        this.init();
    }
    init() {
        this.el = this.tmpl();
        this.container.append(this.el);
        this.bindEvent();
    }
    tmpl() {
        let tm = `<div>
                    <div class="clearfix">
                        <div class="btn-group">
                            <a id="" class="btn btn-primary" href="./index.php?/index/countadd">
                                新建
                                <i class="fa fa-plus"></i>
                            </a>
                            <button style="margin-left:10px;"  class="allDel btn btn-danger">
                                批量删除
                            </button>
                        </div>
                    </div>
                    <table class="display table table-bordered dataTable no-footer" aria-describedby="hidden-table-info_info" role="grid">
                    <thead>
                        <tr role="row">
                            <th>
                                <div class="">
                                    <div style="display: block;
                                        ">
                                        <input type="checkbox" class="allcheck">
                                    </div>
                                </div>
                            </th>
                            <th>ID</th>
                            <th>标题</th>
                            <th>描述</th>
                            <th>时长(单位:秒)</th>
                            <th >创建时间</th>
                            <th>编辑</th>
                        </tr>
                    </thead>

                    <tbody>
                        
                    </tbody>
                </table>
                </div>` ;
        return $(tm);
    }
    uncheckall() {
        this.el.find(".list_checkbox").each(function () {
            let input = $(this);
            input.iCheck("uncheck");
        })
    }
    bindEvent() {
        let self = this;
        let dom = this.el;
        this.table = this.el.find('table').DataTable({
            order: [
                ["5", 'desc']
            ], //按照发布时间降序排序
            page: false,
            searchDelay: 350,
            info: true,
            autoWidth: false,
            searching: true,
            ajax: {
                "url": "./index.php?/countdown/datatable",
            },
            columns: [{
                data: "null",
                sortable: false
            }, {
                data: "id",
                sortable: false
            }, {
                data: "title",
                sortable: false
            }, {
                data: "description",
                sortable: false
            }, {
                data: "duration",
                sortable: false
            }, {
                data: "create_time",
            }, {
                data: "null",
                sortable: false
            }],
            columnDefs: [{
                targets: 0,
                data: null,
                defaultContent: "",
                render: function (e) {
                    let tmp = `<div class="flat-red single-row">
                                    <div style="display: block;
                                        ">
                                        <input type="checkbox" class="list_checkbox">
                                    </div>
                                </div>`;
                    return tmp;
                }
            }, {
                targets: -2,
                data: null,
                defaultContent: "",
                render: function (e) {
                    return utils.crtTime(e * 1000);
                }
            },
            {
                targets: -1,
                data: null,
                defaultContent: "",
                render: function (e) {
                    return `<button class="btn btn-info btn-xs editBtn">编辑</button>
                    <button class="btn btn-danger btn-xs delBtn">删除</button>`;
                }
            }],
            createdRow: function (row, data, index) {
                $(row).data("id", data.id);
                $(row).data("data", data);
                $(row).addClass("countdown_tr");
            },
            "drawCallback": function (set) {
                this.find('.flat-red input').iCheck({
                    checkboxClass: 'icheckbox_flat-red',
                    radioClass: 'iradio_flat-red'
                });
                this.find(".allcheck").iCheck("uncheck");
                self.trigger("draw");
            },
        });
        this.table.on("click", ".editBtn", function (e) {
            let id = $(this).closest("tr").data("id");
            if (!id) return;
            location.href = "./index.php?/index/countedit&id=" + id;
        })
        this.table.on("click", ".delBtn", function (e) {
            let id = $(this).closest("tr").data("id");
            if (!id) return;
            utils.confirm("确定删除?").then(re => {
                $(this).text("删除中...");
                $.ajax({
                    type: "POST",
                    url: "./index.php?/countdown/del",
                    data: { 'id': id },
                    dataType: "jsonp",
                    timeout: "2000"
                }).then(re => {
                    if (re.status == "ok") {
                        self.table.draw(false);
                    } else {
                        utils.alert("网络错误.");
                    }
                }, er => {
                    utils.alert("网络错误.");
                })
            }, er => {

            })
        })
        dom.find(".allcheck").iCheck({
            checkboxClass: 'icheckbox_flat-red',
            radioClass: 'iradio_flat-red'
        })
        let checkflag = true;
        dom.find(".allcheck").on("ifChecked", function (e) {
            checkflag = true;
            dom.find(".list_checkbox").iCheck("check");
        })
        dom.find(".allcheck").on("ifUnchecked", function (e) {
            if (checkflag) {
                dom.find(".list_checkbox").iCheck("uncheck");
            }
        })
        dom.on("ifUnchecked", ".list_checkbox", function () {
            checkflag = false;
            dom.find(".allcheck").iCheck("uncheck");
            let tr = $(this).closest("tr");
            let data = tr.data("data");
            self.trigger("unchecked", data);
        })
        dom.on("ifChecked", ".list_checkbox", function () {
            let tr = $(this).closest("tr");
            let data = tr.data("data");
            self.trigger("checked", data);
        })
        this.el.on('click', ".allDel", function (e) {
            utils.confirm("确定删除?").then(re => {
                let id = [];
                dom.find(".list_checkbox:checked").each((index, val) => {
                    let tr = $(val).closest("tr");
                    let d = tr.data("id");
                    if (d) {
                        id.push(d);
                    }
                })
                let ajaxs = id.map(val => {
                    return $.ajax({
                        type: "post",
                        url: "./index.php?/countdown/del",
                        data: { "id": val },
                        dataType: "jsonp"
                    })
                })
                Promise.all(ajaxs).then(re => {
                    self.table.draw(false);
                }).catch(e => {
                    self.table.draw(false);
                });
            }, er => {

            });
        })
    }
}