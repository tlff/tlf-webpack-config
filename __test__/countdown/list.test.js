import list from "../../src/js/countdown/list";
let app;
beforeEach(() => {
    document.body.innerHTML=`
        <div id="app"></div>
    `;
    app=new list("app");
})
test("test",() => {
    expect(app.el).toMatchSnapshot();
})