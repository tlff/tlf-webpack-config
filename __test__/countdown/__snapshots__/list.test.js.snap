// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 1`] = `
jQuery {
  "0": <div>
    
                    
    <div
      class="clearfix"
    >
      
                        
      <div
        class="btn-group"
      >
        
                            
        <a
          class="btn btn-primary"
          href="./index.php?/index/countadd"
          id=""
        >
          
                                新建
                                
          <i
            class="fa fa-plus"
          />
          
                            
        </a>
        
                            
        <button
          class="allDel btn btn-danger"
          style="margin-left:10px;"
        >
          
                                批量删除
                            
        </button>
        
                        
      </div>
      
                    
    </div>
    
                    
    <div
      class="dataTables_wrapper no-footer"
      id="DataTables_Table_0_wrapper"
    >
      <div
        class="dataTables_length"
        id="DataTables_Table_0_length"
      >
        <label>
          Show 
          <select
            aria-controls="DataTables_Table_0"
            class=""
            name="DataTables_Table_0_length"
          >
            <option
              value="10"
            >
              10
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="100"
            >
              100
            </option>
          </select>
           entries
        </label>
      </div>
      <div
        class="dataTables_filter"
        id="DataTables_Table_0_filter"
      >
        <label>
          Search:
          <input
            aria-controls="DataTables_Table_0"
            class=""
            placeholder=""
            type="search"
          />
        </label>
      </div>
      <table
        aria-describedby="DataTables_Table_0_info"
        class="display table table-bordered dataTable no-footer"
        id="DataTables_Table_0"
        role="grid"
      >
        
                    
        <thead>
          
                        
          <tr
            role="row"
          >
            <th
              aria-label="
                                
                                    <div style=\\"display: block;
                                        \\">
                                        
                                    
                                
                            "
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              
                                
              <div
                class=""
              >
                
                                    
                <div
                  style="display: block;
                                        "
                >
                  
                                        
                  <div
                    class="icheckbox_flat-red"
                  >
                    <input
                      class="allcheck"
                      style="position: absolute; opacity: 0;"
                      type="checkbox"
                    />
                    <ins
                      class="iCheck-helper"
                      style="position: absolute; top: 0%; left: 0%; display: block; width: 100%; height: 100%; margin: 0px; padding: 0px; background: rgb(255, 255, 255); border: 0px; opacity: 0;"
                    />
                  </div>
                  
                                    
                </div>
                
                                
              </div>
              
                            
            </th>
            <th
              aria-label="ID"
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              ID
            </th>
            <th
              aria-label="标题"
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              标题
            </th>
            <th
              aria-label="描述"
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              描述
            </th>
            <th
              aria-label="时长(单位:秒)"
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              时长(单位:秒)
            </th>
            <th
              aria-controls="DataTables_Table_0"
              aria-label="创建时间: activate to sort column ascending"
              aria-sort="descending"
              class="sorting_desc"
              colspan="1"
              rowspan="1"
              tabindex="0"
            >
              创建时间
            </th>
            <th
              aria-label="编辑"
              class="sorting_disabled"
              colspan="1"
              rowspan="1"
            >
              编辑
            </th>
          </tr>
          
                    
        </thead>
        

                    
        <tbody>
          
                        
                    
          <tr
            class="odd"
          >
            <td
              class="dataTables_empty"
              colspan="7"
              valign="top"
            >
              Loading...
            </td>
          </tr>
        </tbody>
        
                
      </table>
      <div
        aria-live="polite"
        class="dataTables_info"
        id="DataTables_Table_0_info"
        role="status"
      >
        Showing 0 to 0 of 0 entries
      </div>
      <div
        class="dataTables_paginate paging_simple_numbers"
        id="DataTables_Table_0_paginate"
      >
        <a
          aria-controls="DataTables_Table_0"
          class="paginate_button previous disabled"
          data-dt-idx="0"
          id="DataTables_Table_0_previous"
          tabindex="0"
        >
          Previous
        </a>
        <span />
        <a
          aria-controls="DataTables_Table_0"
          class="paginate_button next disabled"
          data-dt-idx="1"
          id="DataTables_Table_0_next"
          tabindex="0"
        >
          Next
        </a>
      </div>
    </div>
    
                
  </div>,
  "length": 1,
}
`;
